**Sobre a base de dados**

Os dados de rotatividade de clientes de telecomunicações contêm informações sobre uma empresa de telecomunicações fictícia que forneceu serviços de telefone residencial e Internet para 7.043 clientes na Califórnia.

**Dicionário de dados**


* CustomerId: Customer Id
* Gender: Gender
* SeniorCitizen: Whether the customer is elderly (1.0)
* Partner: Whether the customer has a partner (Yes, No)
* Dependents: Whether the customer has dependents (Yes, No) (child, mother, father...)
* tenure: The number of months the customer stays in the company
* PhoneService: Whether the customer has phone service (Yes, No)
* MultipleLines: Whether the customer has more than one line (Yes, No, No phone service)
* InternetService: Customer's internet service provider (DSL, Fiber optic, No)
* OnlineSecurity: Whether the customer has online security (Yes, No, no Internet service)
* OnlineBackup: Whether the customer has an online backup (Yes, No, no Internet service)
* DeviceProtection: Whether the customer has device protection (Yes, No, no Internet service)
* TechSupport: Whether the customer receives technical support (Yes, No, no Internet service)
* StreamingTV: Whether the customer has broadcast TV (Yes, No, no Internet service)
* StreamingMovies: Whether the customer is streaming movies (Yes, No, no Internet service)
* Contract: Customer's contract period (Month to month, One year, Two years)
* PaperlessBilling: Whether the customer has a paperless invoice (Yes, No)
* PaymentMethod: Customer's payment method (Electronic check, Postal check, Bank transfer, Credit card)
* MonthlyCharges: Amount collected from the customer monthly
* TotalCharges: Total amount collected from the customer
* Churn: Yes = the customer left the company this Month. No = the customer remained with the company.



**Dicionário de dados**


* CustomerId: Customer Id
* Gender: Gênero
* SeniorCitizen: Se o cliente é idoso (1.0)
* Partner: Se o cliente tem um parceiro (Sim, Não)
* Dependents: Se o cliente tem dependentes (Sim, Não) (filho, mãe, pai...)
* tenure: O número de meses que o cliente permanece na empresa
* PhoneService: Se o cliente possui serviço telefônico (Sim, Não)
* MultipleLines: Se o cliente possui mais de uma linha (Sim, Não, Não há serviço telefônico)
* InternetService: Provedor de serviços de Internet do cliente (DSL, Fibra óptica, Não)
* OnlineSecurity: Se o cliente tem segurança online (Sim, Não, sem serviço de Internet)
* OnlineBackup: Se o cliente possui um backup online (Sim, Não, sem serviço de Internet)
* DeviceProtection: Se o cliente possui proteção de dispositivo (Sim, Não, sem serviço de Internet)
* TechSupport: Se o cliente recebe suporte técnico (Sim, Não, sem serviço de Internet)
* StreamingTV: Se o cliente possui TV aberta (Sim, Não, sem serviço de Internet)
* StreamingMovies: Se o cliente está transmitindo filmes (Sim, Não, sem serviço de Internet)
* Contract: Período de contrato do cliente (mês a mês, um ano, dois anos)
* PaperlessBilling: Se o cliente possui uma fatura sem papel (Sim, Não)
* PaymentMethod: Forma de pagamento do cliente (cheque eletrônico, cheque postal, transferência bancária, cartão de crédito)
* MonthlyCharges: Valor arrecadado do cliente mensalmente
* TotalCharges: Valor total arrecadado do cliente
* Churn: Sim = o cliente saiu da empresa este mês. Não = o cliente permaneceu na empresa.

In [2]:
import pandas as pd

path = 'https://raw.githubusercontent.com/profEraylson/datasets/main/churn_telecom.csv'

df = pd.read_csv(path)

df

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 21 columns]

#### Faça uma comparação entre os modelos Regressão Logistica, KNN e Decision Tree e responda as seguintes perguntas:

* Qual modelo apresenta o maior desempenho em treinamento?
* Qual modelo apresenta o maior desempenho em teste?
* Qual modelo você selecionaria para colocar em um sistema de previsão de churn?


Aplique as técnicas vistas em aula e selecione uma ou mais métricas de classificação para avaliar o desempenho dos modelos.


---


*Utilize o seu conhecimento sobre os dados para selecionar e tratar as features da melhor forma.*

*Busque justificar a escolha das técnicas. Exemplo: "Estou aplicando a transformação OneHotEncoder pois...", "Estou utilizando um crossvalidation pois..."*


In [3]:
# Importação das bibliotecas necessárias
from sklearn.model_selection import train_test_split  # Para dividir os dados em conjuntos de treino e teste
from sklearn.preprocessing import OneHotEncoder, StandardScaler  # Para pré-processamento de dados
from sklearn.linear_model import LogisticRegression  # Modelo de Regressão Logística
from sklearn.neighbors import KNeighborsClassifier  # Modelo de k-vizinhos mais próximos (KNN)
from sklearn.tree import DecisionTreeClassifier  # Modelo de Árvore de Decisão
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix  # Para avaliação dos modelos

# Convertendo a coluna "TotalCharges" para float
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')

# Preenchendo os valores ausentes com a média da coluna "TotalCharges"
media_TotalCharges = df['TotalCharges'].mean()
df['TotalCharges'] = df['TotalCharges'].fillna(media_TotalCharges)

# Verificando o tipo da coluna após a conversão
print("Tipo da coluna 'TotalCharges' após a conversão:", df['TotalCharges'].dtype)
print()

# Selecionar features relevantes
X = df.drop(['Churn'], axis=1)  # Features
y = df['Churn']  # Variável alvo

# Resetar índices após a remoção de valores ausentes
X.reset_index(drop=True, inplace=True)  # Redefinindo os índices
y.reset_index(drop=True, inplace=True)

# Converter variáveis categóricas em numéricas usando one-hot encoding
categorical_cols = X.select_dtypes(include=['object']).columns.tolist()  # Identificando colunas categóricas
encoder = OneHotEncoder(drop='first', sparse=False)  # Instanciando o codificador one-hot
X_encoded = encoder.fit_transform(X[categorical_cols])  # Aplicando one-hot encoding
columns_encoded = encoder.get_feature_names_out(categorical_cols)  # Obtendo nomes das novas colunas
X_encoded = pd.DataFrame(X_encoded, columns=columns_encoded)  # Criando DataFrame com novas colunas
X.drop(categorical_cols, axis=1, inplace=True)  # Removendo colunas categóricas originais
X = pd.concat([X, X_encoded], axis=1)  # Concatenando novas colunas com DataFrame original

# Divisão dos dados em treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)  # Dividindo os dados

# Padronizar as features numéricas
scaler = StandardScaler()  # Instanciando o padronizador
X_train_scaled = scaler.fit_transform(X_train)  # Padronizando os dados de treino
X_test_scaled = scaler.transform(X_test)  # Padronizando os dados de teste

# Treinar os modelos
logistic_reg = LogisticRegression(max_iter=1000)  # Instanciando o modelo de Regressão Logística
logistic_reg.fit(X_train_scaled, y_train)  # Treinando o modelo de Regressão Logística

knn = KNeighborsClassifier()  # Instanciando o modelo KNN
knn.fit(X_train_scaled, y_train)  # Treinando o modelo KNN

decision_tree = DecisionTreeClassifier()  # Instanciando o modelo de Árvore de Decisão
decision_tree.fit(X_train_scaled, y_train)  # Treinando o modelo de Árvore de Decisão

# Avaliação dos modelos
models = {'Logistic Regression': logistic_reg,
          'KNN': knn,
          'Decision Tree': decision_tree}  # Dicionário contendo os modelos

for name, model in models.items():
    print(f"Model: {name}")
    train_accuracy = model.score(X_train_scaled, y_train)  # Calculando acurácia nos dados de treino
    test_accuracy = model.score(X_test_scaled, y_test)  # Calculando acurácia nos dados de teste
    print(f"Training Accuracy: {train_accuracy:.2f}")
    print(f"Test Accuracy: {test_accuracy:.2f}")

    y_pred = model.predict(X_test_scaled)  # Realizando previsões nos dados de teste
    print(classification_report(y_test, y_pred))
    print(confusion_matrix(y_test, y_pred))
    print()

Tipo da coluna 'TotalCharges' após a conversão: float64



/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Model: Logistic Regression
Training Accuracy: 1.00
Test Accuracy: 0.79
              precision    recall  f1-score   support

          No       0.79      0.96      0.87      1036
         Yes       0.72      0.31      0.44       373

    accuracy                           0.79      1409
   macro avg       0.76      0.63      0.65      1409
weighted avg       0.78      0.79      0.75      1409

[[992  44]
 [257 116]]

Model: KNN
Training Accuracy: 0.84
Test Accuracy: 0.78
              precision    recall  f1-score   support

          No       0.81      0.91      0.86      1036
         Yes       0.62      0.40      0.48       373

    accuracy                           0.78      1409
   macro avg       0.71      0.65      0.67      1409
weighted avg       0.76      0.78      0.76      1409

[[945  91]
 [225 148]]

Model: Decision Tree
Training Accuracy: 1.00
Test Accuracy: 0.77
              precision    recall  f1-score   support

          No       0.82      0.88      0.85      103

Justificativas:

* Realizei a transformação OneHotEncoder para converter variáveis categóricas em numéricas, pois essas variáveis não podem ser diretamente utilizadas em modelos de machine learning.

* Padronizei as features numéricas para garantir que todas estivessem na mesma escala, o que é importante para modelos como Regressão Logística e KNN que são sensíveis à escala dos dados.

* Utilizei a divisão dos dados em conjunto de treinamento e teste para avaliar o desempenho dos modelos em dados não vistos.

* A avaliação dos modelos foi realizada utilizando métricas de classificação como precisão, recall, F1-score e matriz de confusão para compreender melhor o desempenho dos modelos em relação à previsão de churn.

### Desempenho em Treinamento:
1. **Logistic Regression:** Acurácia de treinamento de 100%.
2. **KNN:** Acurácia de treinamento de 84%.
3. **Decision Tree:** Acurácia de treinamento de 100%.

### Desempenho em Teste:
1. **Logistic Regression:** Acurácia de teste de 78%.
2. **KNN:** Acurácia de teste de 77%.
3. **Decision Tree:** Acurácia de teste de 75%.

### Escolha para um Sistema de Previsão de Churn:
Considerando os resultados acima, a Logistic Regression apresentou a maior acurácia tanto no treinamento quanto no teste. Além disso, é importante notar que a Logistic Regression possui uma precisão superior na previsão de clientes que cancelam o serviço (classe "Yes"), o que pode ser crucial para um sistema de previsão de churn, pois minimizar falsos negativos (clientes que cancelam, mas foram classificados como não cancelantes) é fundamental para tomar medidas preventivas. Portanto, com base nas métricas fornecidas, a Logistic Regression seria a escolha recomendada para um sistema de previsão de churn.